### PR TITLE
Proposed parents configuration file to be used with remap.config

### PR DESCRIPTION
**Pull request by jrushford**
This PR is to add an example parents.yaml file that eventually will be used with remap.config.  A new tag in remap say, @parent=parents.yaml may be used to configure parent/origin next hop on a per remap basis.

Looking for comments on this file.

### CODE REVIEW

File: proxy/config/parents.yaml.example [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          alternate_rings: # optional
+            response_codes:
+              - 522
+            health_check:
+              - active
+
+##
+# hosts.yaml
+hosts:
+  p1: # shorthand name of host object
```
Comment (bryancall) [inline proxy/config/parents.yaml.example]: Do you need to have a shorthand name?  If not then make the host a list.

Comment (bryancall) [reply proxy/config/parents.yaml.example]: Looks like yes - Looking at primary_ring above.

**Comment by zwoop** [comment]
We had some discussions yesterday, so this proposal will be updated from that (soon hopefully)

**Comment by vmamidi** [comment]
Here is the updated Parent configuration.
[parent.yml.txt](https://github.com/apache/trafficserver/files/2168128/parent.yml.txt)

File: proxy/config/parents.yaml.example [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+map / http://bob parents=bob.yaml
+map /foo http://bob parents=foo.yaml
+
+##########
+## bob.yaml
+include: [hosts.yaml] # Optional, but must specify hosts somehow
+
+strategy:
+    protocol: h2 # allowed protocol. optional by default uses the remapped protocol
+    name: # Name of the strategy: Enum of 'consistent_hash' or 'first_live' or 'rr_strict' or 'rr_ip' or 'latched'
```
[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/config/parents.yaml.example]: this is a list of the supported selection strategies that we know about so far.  Others can be added.  Is that what you're asking @SolidWallOfCode

File: proxy/config/parents.yaml.example [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    hash_key: # key to use for Hashing. Enum of 'url' or 'uri' or 'hostname' or 'path' or 'path+query' or 'cache_key' or 'path+fragment'
+    primary_ring: # required: list of servers and their weights
+      weights: # optional: weights used for the servers
+        p1: 1.0
+        p2: 2.0
+      hosts: # required list of servers
```
[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/config/parents.yaml.example]: In our discussion we talked about defining all the hosts and their attributes in a separate file that could imported. Then in the strategy section which would be for a particular remap, you could choose the set of hosts you want for this remap and then choose how they should be weighted.

**Comment by vmamidi** [comment]
@bryancall and @SolidWallOfCode can you guys have a look at the new file?

**Comment by zwoop** [comment]
Yeh, I like this new format. In particular, it lets someone minimize duplications, by using anchors for shared pieces, and appropriate inheritance. Like, one could take the above one step further, and isolating the protocol specifications, healthchecks etc. as an anchored resource, and then include that. Or, one could exclude the "host" portion from those anchored sections, and override that locally in the policy sections. It's a lot nicer than all the duplications that we had initially.

**Comment by SolidWallOfCode** [comment]
It does bring Bryan's question back - what is the use of the short name in the primary and secondary ring definitions?

**Comment by SolidWallOfCode** [comment]
"it allows you to avoid repeating all the unchanging data about a host in every strategy file. For instance, imagine"  - but as Leif points out, YAML does this with anchors and references, as is done in the latest example configuration file. In that file, what purpose does "p1" in the "strategy" section serve? When would it be referenced instead of "*anchornameforp1"? Contrariwise, in your description, what purpose doe the "hosts" table serve? It seems to be me one is useful in the manner you describe, but I don't see why both are.

**Comment by jrushford** [comment]
I added an '#include' to the example to show that a strategy yaml file, bar.yaml, could use '#include our hosts.yaml'. so that the two files are concatenated into one yaml document.  In the example yaml parser I've started on, I've implemented this.

**Comment by SolidWallOfCode** [comment]
Will this be updated in accordance with the L7R Working Group meeting last week?

**Comment by jrushford** [comment]
Alan, I’ll update that today.

**Comment by jrushford** [comment]
It’s been updated, see #3870

**Comment by vmamidi** [comment]
Primary and secondary group names give an impression that there are only two groups but there is no reason not to have multiple groups and transfer from group to another can happen based on failover. 
For example, 502 can result in moving to g3 whereas 404 can result in moving g2 etc..

**Comment by mlibbey** [comment]
Sure -- just wondering how we'd configure the consistent hash feature that tells ATS to choose from group 2 if the specific chosen host in group 1 is down (as opposed to choosing another host in group 1).

**Comment by jrushford** [comment]
good point @mlibbey, I don't see anyway to configure that so that primary/secondary order either and what should the behavior be if there are three groups?  Also, how are the multiple groups used with the round robin strategies?  Should another weight or order be added to the strategy group list as in this:

`strategy:
   policy: # Selection strategy policy: Enum of 'consistent_hash' or 'first_live' or 'rr_strict' or 'rr_ip' or 'latched'
   hash_key: # optional key to use for Hashing. Enum of 'url' or 'uri' or 'hostname' or 'path' or 'path+query' or 'cache_key' or 'path+fragment'
   groups:
     - <<: *g1
        choice: 1
     - <<: *g2
        choice: 2
`

**Comment by SolidWallOfCode** [comment]
This is one of the reasons I suggested a recursive structure where a ring could have as elements other rings. Then it's easy - you define ring A to be a linear fail over of ring B (the primary pod) and ring C (the secondary).

I think the structure suggested in the working group, you would have ring B and ring C, and then a strategy that said "try B, then failover to C".

**Comment by SolidWallOfCode** [comment]
@mlibbey Isn't g1 primary and g2 secondary because it's a list and g1 is first?

**Comment by jrushford** [comment]
using the order in which the groups are in the list makes sense ie, g1 is primary because it's first.

**Comment by jrushford** [comment]
there are two keys, in the example, "health_check" and "healthcheck".  they have different meanings where in a 'host' context it refers to a health check url and then in strategy failover context it means either an 'active' or 'passive' health check.  Is there any reason they can't be spelled the same ie, 'healthcheck' or 'health_check' to avoid confusion on which spelling is used in the differing contexts or should one be renamed entirely?  Maybe just use combine 'healthcheck' and 'url' in the host context to 'healcheck_url'?

**Comment by jrushford** [comment]
Is the proposed format now show okay with everyone?

**Comment by bryancall** [comment]
[approve ci fedora]

**Comment by bryancall** [comment]
[approve ci rat]

**Comment by jrushford** [comment]
@bryancall, @zwoop, @vmamidi, @SolidWallOfCode I don't think we need this PR any-longer and it should be closed provided everyone is okay with #5960 and the new format of strategies.yaml

**Comment by bryancall** [comment]
@jrushford sounds good

**Pull request closed without merging**